### ISSUE TITLE
Reviewable Enterprise doesn't work for our Chinese colleagues due to directly talking to Firebase

### ISSUE THREAD

**Issue by schreter** [Reviewable/Reviewable#1128]
We've got some new colleagues on our team, but they are located in China. Unfortunately, there is no direct access to Firebase from China (since it's on Google servers) and thus we can't get them to work with Reviewable, disrupting our workflow.

They do have full network access to our GHE instance and also to our Reviewable instance (since they are both inside of our company network), but it still doesn't work, since the browser seems to directly talks to Firebase (which is blocked by the great firewall).

It would be very helpful to have a possibility to proxy Firebase requests through Reviewable instance running in a zone where Firebase is accessible.

Or, do you have other ideas how we can enable our Chinese colleagues work with Reviewable?

Thanks in advance.

**Comment by pkaminski** [comment]
I'll do my best to help out, but please note that in general any workarounds for the Great Firewall are temporary at best so we cannot guarantee that Reviewable will be usable from China.

In principle, your best bet would be to use a VPN through your corporate network so that your colleagues' traffic would exit in Europe, but my understanding is that this is unlikely to work well in China and may even incur regulatory trouble.  Still, it might be worth checking with your colleagues for ideas along those lines — perhaps there's an authorized way to make this work.

The next option would be to set up a proxy on your corporate network and set `REVIEWABLE_FIREBASE_URL` to the proxy's address.  Feedback on accessing Firebase RTDB via a proxy is mixed and this doesn't appear to be officially supported by the JS SDK but it's worth a try.  Ideally, the proxy should support upgrading the connection to a web socket, but if it doesn't then (in theory) the SDK should fall back to HTTP long polling.

If this approach works then we could add a way for your colleagues to override the `REVIEWABLE_FIREBASE_URL` in their clients so everyone else doesn't have to go through the proxy.  If long polling is necessary then there's also an internal API to force it that we could expose to your colleagues to make establishing the connection faster.  We probably wouldn't want to integrate the proxy proper into Reviewable server's, though, as that's a responsibility and maintenance burden we'd rather not take on.

If the proxy doesn't work then I'm out of ideas for other options to try.  Having the client directly access Firebase is a core tenet of Reviewable's architecture and not something we'll be able to change in the foreseeable future.

**Comment by schreter** [comment]
Well, the proxy was also my first idea, but it seems like Firebase redirects from the URL to another one after connecting via 3xx code. Thus, a single proxy won't do. Probably we'll need multiple and do some DNS hacks as well.

Also, we're not circumventing the Great Firewall :-). The colleagues have official access to our corporate network via VPN peering, but only to our corporate network, not to the outside. So having a proxy *within* Reviewable would be likely the most conformant way of doing things.

Anyway, I'll try with the colleagues next ween to somehow get it working. It would be a shame should they not be able to use Reviewable (that would disrupt our workflow).

BTW, I'm wondering, how other Chinese people who directly work on open-source GitHub projects can use Reviewable from China. There, it seems to work w/o problems. Do you have some special handling there?

**Comment by pkaminski** [comment]
> Well, the proxy was also my first idea, but it seems like Firebase redirects from the URL to another one after connecting via 3xx code.

That's the way it used to work but I don't see it happening any more on our SaaS production instance.  Instead, if the `firebase:host:reviewable.firebaseio.com` key is missing from local storage (substitute your Firebase URL), I see it making a request to the nominal URL that gets a direct 101 Switching Protocols response.  (It then passes the actual URL via the websocket and sets it in local storage so it can connect directly next time.)  Perhaps they upgraded their North American instances recently and will make the change to European ones later?

In any case, in practice the final `firebaseio.com` URL is pretty stable, so I think you could set up a proxy directly to that and bypass the redirect altogether.  If you provide the proxy's address as `REVIEWABLE_FIREBASE_URL` you shouldn't need to mess with DNS either.  (Older SDKs crashed on non-Google URLs, but I believe modern ones will tolerate them.)

> BTW, I'm wondering, how other Chinese people who directly work on open-source GitHub projects can use Reviewable from China. There, it seems to work w/o problems. Do you have some special handling there?

We don't do anything special to accommodate mainland China users and I'm not personally aware of any who use Reviewable, though it's certainly possible that some implemented their own workarounds.  The only other explanation I can think of is that somebody added an explicit bypass for Reviewable to the Great Firewall without telling us, but that seems rather implausible.  😂

**Comment by pkaminski** [comment]
Hey @schreter, Happy New Year!  Did you manage to get a proxy working?

**Comment by pkaminski** [comment]
Closing, but happy to reopen if you have further questions or ideas.
